<!DOCTYPE html>
<html lang="ja">
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="description" content="" />
    <meta name="author" content="Python会" />
    <meta name="generator" content="Pelican (VoidyBootstrap theme)" />
    <meta property="og:type" content="article"/>
    <meta property="og:title" content="がんゲノム変異シグネチャー解析"/>
    <meta property="og:url" content="../../../blog/2019/06/cancer_signature.html"/>
      <meta property="article:section" content="技術ブログ"/>
        <meta property="article:tag" content="Bioinformatics"/>
      <meta property="article:author"
            content="Python会"/>
        <meta property="og:image"
              content="../../../images/images/logo.jpg"/>
    <meta name="twitter:card" content="summary"> 
    <meta name="twitter:title" content="がんゲノム変異シグネチャー解析">
    <meta name="twitter:description" content="">
    <meta name="twitter:site" content="@oumed_python">
    <meta name="twitter:creator" content="@oumed_python">
    <meta name="twitter:domain" content="../../..">
        <meta property="twitter:image"
              content="../../../images/images/logo.jpg"/>

    <title>がんゲノム変異シグネチャー解析 - 大阪大学医学部 Python会 <span style="font-size:smaller;">(情報医科学研究会)</span></title>

   
        <link rel="stylesheet"
              href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css"
              integrity="sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk"
              crossorigin="anonymous">
      <link rel="stylesheet"
            href="https://use.fontawesome.com/releases/v5.13.0/css/all.css"
            integrity="sha384-Bfad6CLCknfcloXFOyFnlgtENryhrpZCe29RTifKEixXQZ38WheV+i/6YWSzkz3V"
            crossorigin="anonymous">


      <link rel="stylesheet" href="../../../theme/css/pygment.css" />
      <link rel="stylesheet" href="../../../theme/css/theme.css" />
      <link rel="stylesheet" href="../../../theme/css/voidybootstrap-custom.css" />

    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js" integrity="sha384-FFgGfda92tXC8nCNOxrCQ3R8x1TNkMFqDZVQdDaaJiiVbjkPBXIJBx0o7ETjy8Bh" crossorigin="anonymous"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js" integrity="sha384-ZoaMbDF+4LeFxg6WdScQ9nnR1QC2MIRxA1O9KWEXQwns1G8UNyIEZIQidzb0T1fo" crossorigin="anonymous"></script>
    <![endif]-->

    <link rel="shortcut icon" href="../../../favicon.ico" />
    <meta name="siteurl" property="og:siteurl" content="../../..">
    <meta name="source-repository" property="og:source-repository" content="https://github.com/oumpy/hp_management.git">
        <link href="https://oumpy.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="大阪大学医学部 Python会 (情報医科学研究会) Atom Feed" />
        <link href="https://oumpy.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="大阪大学医学部 Python会 (情報医科学研究会) RSS Feed" />
  </head>

  <body>
   
    <nav class="navbar navbar-expand py-0 fixed-top">
      <div class="container">
      <div class="row dropdown-hover-all">
        <a class="navbar-brand" href="../../../" rel="home">
          <i class="fa fa-home fa-fw fa-lg text-muted"> </i>
          </a>
		<button type="button" class="navbar-toggler" 
            data-toggle="collapse" data-target="#main-navbar-collapse"
            area-controls="main-navbar-collapse" area-extended="false">
      <span class="navbar-toggler-icon"></span>
		  <span class="sr-only">Toggle navigation</span>
		  <span class="icon-bar"></span>
		  <span class="icon-bar"></span>
		  <span class="icon-bar"></span>
		</button>
      <div class="collapse navbar-collapse" id="main-navbar-collapse">
        <div class="dropdown-hover-all">
          <ul class="nav navbar-nav mr-auto">
            <li class="nav-item dropdown dropdown-hover">
<a class="nav-link dropdown-toggle" id="dropdown1" href="../../../index.html" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">Python会について</a>
<div class="dropdown-menu" aria-labelledby="dropdown1">
<a class="dropdown-item" href="../../../index.html">Python会について</a>
<div class="dropdown-divider"></div>
<a class="dropdown-item" href="../../../constitution.html">会規約</a>
</div>
</li>
<li class="nav-item dropdown dropdown-hover">
<a class="nav-link dropdown-toggle" id="dropdown4" href="../../../activities.html" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">活動内容</a>
<div class="dropdown-menu" aria-labelledby="dropdown4">
<a class="dropdown-item" href="../../../activities.html">活動内容</a>
<div class="dropdown-divider"></div>
<a class="dropdown-item" href="../../../activities.html#pelican-subsections-0">学習・研究</a>
<a class="dropdown-item" href="../../../activities.html#pelican-subsections-4">ワークショップ等の開催</a>
<a class="dropdown-item" href="../../../activities.html#pelican-subsections-5">医学科学生用計算サーバの管理運用</a>
<a class="dropdown-item" href="../../../activities.html#pelican-subsections-6">過去の活動</a>
</div>
</li>
<li class="nav-item dropdown dropdown-hover">
<a class="nav-link dropdown-toggle" id="dropdown10" href="../../../achievements.html" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">実績</a>
<div class="dropdown-menu" aria-labelledby="dropdown10">
<a class="dropdown-item" href="../../../achievements.html">実績</a>
<div class="dropdown-divider"></div>
<a class="dropdown-item" href="../../../achievements.html#pelican-subsections-0">勉強会・成果発表活動</a>
<a class="dropdown-item" href="../../../achievements.html#pelican-subsections-4">会外部との連携によるイベント開催</a>
<a class="dropdown-item" href="../../../achievements.html#pelican-subsections-7">コンペティション実績</a>
<a class="dropdown-item" href="../../../achievements.html#pelican-subsections-12">研究開発活動</a>
<a class="dropdown-item" href="../../../achievements.html#pelican-subsections-16">所属メンバー論文</a>
<a class="dropdown-item" href="../../../achievements.html#pelican-subsections-20">出版記事等</a>
</div>
</li>
<li class="nav-item dropdown dropdown-hover active">
<a class="nav-link dropdown-toggle" id="dropdown18" href="../../../blog.html" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">技術ブログ</a>
<div class="dropdown-menu" aria-labelledby="dropdown18">
<a class="dropdown-item active" href="../../../blog.html">技術ブログ</a>
<div class="dropdown-divider"></div>
<a class="dropdown-item" href="../../../authors.html">著者一覧</a>
<a class="dropdown-item" href="../../../archives.html">過去記事一覧</a>
</div>
</li>
<li class="nav-item">
<a class="nav-link" href="../../../news.html">お知らせ</a>
</li>
<li class="nav-item dropdown dropdown-hover">
<a class="nav-link dropdown-toggle" id="dropdown23" href="../../../recruit.html" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">会員募集</a>
<div class="dropdown-menu" aria-labelledby="dropdown23">
<a class="dropdown-item" href="../../../recruit.html">会員募集</a>
<div class="dropdown-divider"></div>
<a class="dropdown-item" href="../../../recruit.html#pelican-subsections-0">新歓2020年度</a>
</div>
</li>
<li class="nav-item">
<a class="nav-link" href="../../../contact.html">Contact</a>
</li>
        </ul> <!-- /nav -->
        </div> <!-- /dropdown-hover -->
      </div> <!-- /navbar-collapse -->
    </div> <!-- /row -->
    </div> <!-- /container -->
    </nav> <!-- /navbar -->

	<div class="jumbotron" id="overview">
	  <div class="container">
		<h1><a href="../../../">大阪大学医学部 Python会 <span style="font-size:smaller;">(情報医科学研究会)</span></a></h1>
		<p class="lead">Now is better than never.</p>
	  </div>
	</div>
    <meta name="article:slug" property="og:article:slug" content="cancer_signature">
    <meta name="article:category" property="og:article:category" content="技術ブログ">
    <meta name="article:category:slug" property="og:article:category:slug" content="blog">
    <meta name="article:date" property="og:article:date" content="2019-06-21">

    <div class="container" id="main-container">
      <div class="row">
        <div class="col-md-9" id="content">
<article itemscope="itemscope" itemtype="http://schema.org/BlogPosting">
  <header class="article-header">
<h1>
  <a href="../../../blog/2019/06/cancer_signature.html" rel="bookmark"
     title="Permalink to がんゲノム変異シグネチャー解析">
    がんゲノム変異シグネチャー解析
  </a>
</h1><div class="article-header-info">
  <p>
<abbr class="article-header-date">
  2019-06-21(Fri)
</abbr>       - Posted by 
<a href="../../../author/shan-tian.html">山田</a>    in 
    <a href="../../../blog.html">
      技術ブログ</a>
    &nbsp;&nbsp;
        <a href="../../../tag/bioinformatics.html"><img class="notebook-badge-image" src="https://img.shields.io/static/v1?label=Tag&message=Bioinformatics&labelColor=black&color=darkorange" alt="tag:Bioinformatics" title="tag:Bioinformatics"></img></a>
  </p>
</div> <!-- /.article-header-info --><div id="colablink"></div><div class="toc_box">
    <span class="box-title">Contents</span>
      <ul id="toc"></ul>
</div>  </header>
  <div class="content-body" itemprop="text articleBody">
	<p>がんゲノムの変異シグネチャーの推定にチャレンジしてみました。</p>
<p><img src="../../../blog/2019/06/images/cancer_signature_figs/1.png" width="400px"></p>
<h2><span id="pelican-subsections-0">きっかけ</span></h2>
<p>先日の遺伝学の講義にて、がんゲノムの変異シグネチャーを教えていただき、やり方も詳しく書かれていたので、挑戦してみました。</p>
<p>講義ではRでしたので、Pythonでも同じ結果が得られるか、Pythonの勉強も兼ねての挑戦記録です。</p>
<h2><span id="pelican-subsections-1">がんゲノムの変異シグネチャーとは？？</span></h2>
<p><img alt="2" src="../../../blog/2019/06/images/cancer_signature_figs/2.png"></p>
<p>（がんゲノム情報学　中谷先生のスライドより）</p>
<h2><span id="pelican-subsections-2">トピックモデルとは？？</span></h2>
<p><img alt="3" src="../../../blog/2019/06/images/cancer_signature_figs/3.png"></p>
<p>（<a href="https://www.albert2005.co.jp/knowledge/machine_learning/topic_model/about_topic_model">https://www.albert2005.co.jp/knowledge/machine_learning/topic_model/about_topic_model</a>より）</p>
<h3><span id="pelican-subsections-3">がんゲノムの場合のトピックは？？</span></h3>
<p><img alt="4" src="../../../blog/2019/06/images/cancer_signature_figs/4.png"></p>
<p>（がんゲノム情報学　中谷先生のスライドより）</p>
<p>一塩基ごとの変異情報がまとめられたテキストがDocumentsとして、がんの原因となる要素を一つ一つの「topic」、変異シグネチャーとして、トピックモデルを用いて推定します。</p>
<h2><span id="pelican-subsections-4">解析の方法</span></h2>
<p><img alt="5" src="../../../blog/2019/06/images/cancer_signature_figs/5.png"></p>
<p>（がんゲノム情報学　中谷先生のスライドより）</p>
<p>データベースにある変異データには、<br>
ゲノムの位置、その位置における変異前の塩基配列、変異後の配列<br>
の情報が書かれています。</p>
<p>そこから変異の情報を、C&gt;T_AGのような形でテキストファイルにIDごとに記録していきます。<br>
鋳型鎖と相補鎖の関係より、<strong>変異前がCかTの側のみ</strong> を集計します。（＊注1）</p>
<p>その後、各IDごとに変異データのテキストファイルを集め、変異情報をDocumentsとして、トピック、変異シグネチャーを推定していきます。</p>
<h2><span id="pelican-subsections-5">使用したツール</span></h2>
<ol>
<li>pandas  </li>
<li>
<p>matplotlib<br>
Pythonでおなじみのツールですね。</p>
</li>
<li>
<p>biopython<br>
Python 用の計算機分子生物学パッケージです。<br>
核酸やアミノ酸の配列情報を扱うためのFASTAファイルの処理が簡単になります。<br>
（ドキュメントは読み切れていないので、いずれ読みたいです…）<br>
公式：<a href="https://biopython.org/">https://biopython.org/</a><br>
参考：</p>
</li>
<li><a href="https://qiita.com/sato32ha6/items/6b762b0d0314a5db7dc3">https://qiita.com/sato32ha6/items/6b762b0d0314a5db7dc3</a></li>
<li>
<p><a href="https://bi.biopapyrus.jp/python/biopython/seq.html">https://bi.biopapyrus.jp/python/biopython/seq.html</a></p>
</li>
<li>
<p>gensim（topicmodel）<br>
今回の核となるトピックモデルについてはこちらを使います。<br>
gensimは、様々なトピックモデルを実装したPythonライブラリです。<br>
"<em>topic modeling for humans</em>"とあるように、実装が大変なトピックモデルを簡単に使うことができます。<br>
今回はその中でも、LDA（Latent Dirichlet Allocation）を用います。<br>
参考：</p>
</li>
<li><a href="https://www.sejuku.net/blog/67863">https://www.sejuku.net/blog/67863</a><br>
gensimの使い方についてはこれが一番わかりやすかったです）　　　</li>
<li>
<p><a href="https://www.machinelearningplus.com/nlp/topic-modeling-gensim-python/">https://www.machinelearningplus.com/nlp/topic-modeling-gensim-python/</a></p>
</li>
<li>
<p>wordcloud<br>
matplotlib を利用して Word Cloud を作成できる Python ライブラリです。<br>
wordcloud自体は、文章中で出現頻度が高い単語を複数選び出し、その頻度に応じた大きさで図示する手法です。</p>
</li>
</ol>
<h2><span id="pelican-subsections-6">実践　〜解析〜</span></h2>
<h3><span id="pelican-subsections-7">COSMICとは？</span></h3>
<p>（HP：https://cancer.sanger.ac.uk/cosmic/）</p>
<blockquote>
<p>COSMIC, the Catalogue Of Somatic Mutations In Cancer, is the world’s largest and most comprehensive resource for exploring the impact of somatic mutations in human cancer.</p>
</blockquote>
<p>（公式サイトより）</p>
<p>がんと関連する体細胞変異の情報を集積したデータベースです。</p>
<p>今回はこちらにある変異データを解析の対象としています。</p>
<p>参考：統合TV　<a href="https://togotv.dbcls.jp/20180127.html">https://togotv.dbcls.jp/20180127.html</a></p>
<h3><span id="pelican-subsections-8">データのダウンロード</span></h3>
<p>がんゲノムの変異データをCOSMCのサイトからダウンロードします。
- <a href="https://cancer.sanger.ac.uk/cosmic/download">https://cancer.sanger.ac.uk/cosmic/download</a></p>
<p>データを無料でダウンロードするには、学生としての登録が必要でした。
今回は様々なデータの中でも、Non coding variantsのものを使いました。
<img alt="6" src="../../../blog/2019/06/images/cancer_signature_figs/6.png"></p>
<h3><span id="pelican-subsections-9">リファレンス配列の入手</span></h3>
<p>今回はUCSCからhg38をダウンロードしました。</p>
<div class="highlight"><pre><span></span><code>$ wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
</code></pre></div>


<p>こちらのフォーマットはfastaで、</p>
<p><img alt="7" src="../../../blog/2019/06/images/cancer_signature_figs/7.png"></p>
<p>のような感じで並んでいます。</p>
<p>配列にAやaなど、大文字と小文字が並んでいるのは、大文字の A, C, G, T はコード領域、小文字の a, c, g, t はコード領域以外を表します。</p>
<h3><span id="pelican-subsections-10">変異データを覗いてみる</span></h3>
<p>COSMICから入手したデータがどんなフォーマットなのか見てみました。</p>
<p><img alt="8" src="../../../blog/2019/06/images/cancer_signature_figs/8.png"></p>
<p>ここからサンプルID、ゲノムのposition、元の配列、変異後の配列、という形で抽出し手確認しました。</p>
<p><img alt="9" src="../../../blog/2019/06/images/cancer_signature_figs/9.png"></p>
<p>公式サイトによると、変異データのフォーマットは以下のようになっているようです。</p>
<blockquote>
<p><strong>[column number:label] Heading</strong><br>
<strong>[1:A] Sample name,Sample id,Tumour id</strong> – A sample is an instance of a portion of a tumour being examined for mutations. The sample name can be derived from a number of sources. In many cases it originates from the cell line name. Other sources include names assigned by the annotators, or an incremented number assigned during an anonymisation process. A number of samples can be taken from a single tumour and a number of tumours can be obtained from one individual. A sample id is used to identify a sample within the COSMIC database. There can be multiple ids, if the same sample has been entered into the database multiple times from different papers.<br>
<strong>[4:D] Primary Site</strong> – The primary tissue/cancer from which the sample originated. More details on the tissue classification are avaliable from <a href="https://cancer.sanger.ac.uk/cell_lines/classification">here</a>. In COSMIC we have standard classification system for tissue types and sub types because they vary a lot between different papers.<br>
<strong>[5:E] Site Subtype 1</strong> – Further sub classification (level 1) of the samples tissue of origin.<br>
<strong>[6:F] Site Subtype 2</strong> – Further sub classification (level 2) of the samples tissue of origin.<br>
<strong>[7:G] Site Subtype 3</strong> – Further sub classification (level 3) of the samples tissue of origin.<br>
<strong>[8:H] Primary Histology</strong> – The histological classification of the sample.<br>
<strong>[9:I] Histology Subtype 1</strong> – Further histological classification (level 1) of the sample.<br>
<strong>[10:J] Histology Subtype 2</strong> – Further histological classification (level 2) of the sample.<br>
<strong>[11:K] Histology Subtype 3</strong> – Further histological classification (level 3) of the sample.<br>
<strong>[12:L] Mutation Id</strong> – unique mutation identifier.<br>
<strong>[13:M] Zygosity</strong> – Information on whether the mutation was reported to be homozygous , heterozygous or unknown within the sample.<br>
<strong>[14:N] GRCh</strong> – The coordinate system used –<br>
38 = GRCh38/Hg38<br>
37 = GRCh37/Hg19<br>
<strong>[15:O] Genome position</strong> – The genomic cooridnate of the mutation.<br>
<strong>[16:P] Mutation somatic status</strong> – Information on whether the sample was reported to be Confirmed Somatic, Previously Reported or Variant of unknown origin –<br>
Confirmed Somatic = if the mutation has been confimed to be somatic in the experiment by sequencing both the tumour and a matched normal from the same patient.<br>
variant of unknown origin = when the mutation is known to be somatic but the tumour was sequenced without a matched normal.<br>
Previously observed = when the mutation has been reported as somatic previously but not in current paper.<br>
<strong>[17:Q] WT SEQ</strong> – wild type sequence.<br>
<strong>[18:R] MUT SEQ</strong> – Mutated sequence.<br>
<strong>[19:S] SNP</strong> – All the known SNPs are flagged as ‘y’ defined by the 1000 genomes project, dbSNP and a panel of 378 normal (non-cancer) samples from Sanger CGP sequencing.<br>
<strong>[20:T] FATHMM_MKL_NON_CODING_SCORE</strong> – FATHMM-MKL non-coding score. A p-value ranging from 0 to 1 where &gt;= 0.7 is functionally significant.<br>
<strong>[21:U] FATHMM_MKL_NON_CODING_GROUPS</strong> – FATHMM-MKL group classification. More details from <a href="https://cancer.sanger.ac.uk/cosmic/analyses">here</a>.<br>
<strong>[22:V] FATHMM_MKL_CODING_SCORE</strong> – FATHMM-MKL coding score (p-value ranging from 0 to 1).<br>
<strong>[23:W] FATHMM_MKL_CODING_GROUPS</strong> – FATHMM-MKL group classification (coding). More details from <a href="https://cancer.sanger.ac.uk/cosmic/analyses">here</a>.<br>
<strong>[24:X] Whole Genome Reseq</strong> – if the enitre genome is sequenced.<br>
<strong>[25:Y] Whole_Exome</strong> – if the enitre exome is sequenced.<br>
<strong>[26:Z] Id Study</strong> – Lists the unique Ids of studies that have involved this non coding mutation.<br>
<strong>[27:AA] Pubmed_PMID</strong> – The PUBMED ID for the paper that the sample was noted in.</p>
</blockquote>
<h3><span id="pelican-subsections-11">変異ファイルの処理</span></h3>
<p>ここが一番苦労しました。。。</p>
<h4><span id="pelican-subsections-12">簡単な流れ</span></h4>
<ol>
<li>変異データをpandasで読み込む</li>
<li>MUT_SEQについて、genome_positionからリファレンス配列を参照して、変異配列の前後の配列を探し出す</li>
<li>DataFrameの新しいcolumnとして配列を追加していく</li>
<li>ID_tumorごとに、配列の変異をテキストファイルに書き出していく。</li>
<li>tumorごとの変異についてのテキストファイルをもとに、トピックモデルをgensimを使って作成する。</li>
<li>可視化する。</li>
</ol>
<p>以下実際のコードです。</p>
<h4><span id="pelican-subsections-13">ライブラリのimport</span></h4>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
</code></pre></div>


<h4><span id="pelican-subsections-14">COSMICの変異データをpandasでDataFrameとして読み込みます。</span></h4>
<div class="highlight"><pre><span></span><code><span class="n">mut_file</span><span class="o">=</span><span class="s1">&#39;CosmicNCV.tsv&#39;</span>
<span class="n">ref_fasta</span><span class="o">=</span><span class="s1">&#39;hg38.fa&#39;</span><span class="n">df_cosmic</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">mut_file</span><span class="p">)</span>
<span class="n">df_cut</span><span class="o">=</span><span class="n">df_cosmic</span><span class="p">[[</span><span class="s1">&#39;ID_SAMPLE&#39;</span><span class="p">,</span> <span class="s1">&#39;ID_tumour&#39;</span><span class="p">,</span><span class="s1">&#39;genome position&#39;</span><span class="p">,</span><span class="s1">&#39;WT_SEQ&#39;</span><span class="p">,</span> <span class="s1">&#39;MUT_SEQ&#39;</span><span class="p">]]</span>
<span class="n">df_cut</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>


<p><img alt="10" src="../../../blog/2019/06/images/cancer_signature_figs/10.png"></p>
<p>上記のようにDataFrameとしてきれいに読み込めました。</p>
<h4><span id="pelican-subsections-15">リファレンス配列の読み込み</span></h4>
<p>ここではbiopythonを使って簡単に配列部分を取り出します。</p>
<div class="highlight"><pre><span></span><code><span class="c1">#リファレンスファイルについて</span>
<span class="n">list_id</span><span class="o">=</span><span class="p">[]</span>
<span class="n">list_desc</span><span class="o">=</span><span class="p">[]</span>
<span class="n">list_seq</span><span class="o">=</span><span class="p">[]</span><span class="c1">#biopythonを用いて、idやdesc、seqを抽出</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">ref_fasta</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
<span class="n">id_part</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>
<span class="n">desc_part</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">description</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span>

<span class="n">list_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">id_part</span><span class="p">)</span>
<span class="n">list_desc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">desc_part</span><span class="p">)</span>
<span class="n">list_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</code></pre></div>


<p>塩基配列はlist_seqに入っています。</p>
<p><img alt="11" src="../../../blog/2019/06/images/cancer_signature_figs/11.png"></p>
<p>全てリスト形式なので、</p>
<div class="highlight"><pre><span></span><code><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">list_seq</span><span class="p">[</span><span class="mi">49</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">list_desc</span><span class="p">[</span><span class="mi">49</span><span class="p">])</span><span class="mi">19829559</span>

<span class="n">chr3</span>
</code></pre></div>


<p>のように各染色体ごとに見るこができます。</p>
<h4><span id="pelican-subsections-16">COSMICのdf（DataFrame）の操作</span></h4>
<ol>
<li>df_cutのgenome_positionについて、染色体番号とゲノムの位置に分割して新しいカラムとして追加します。</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1">#chr、genome position用の空のリストを作成</span>

<span class="n">list_cut_chr</span><span class="o">=</span><span class="p">[]</span>
<span class="n">list_cut_pos</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df_cut</span><span class="p">)):</span>
<span class="n">list_cut_chr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;[:-]&#39;</span><span class="p">,</span><span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;genome position&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]))[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">list_cut_pos</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;[:-]&#39;</span><span class="p">,</span><span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;genome position&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]))[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1">#df_cutに新しいカラムとして追加。</span>

<span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">list_cut_chr</span><span class="p">)</span>

<span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;single position&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">list_cut_pos</span><span class="p">)</span>
<span class="n">df_cut</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>


<p><img alt="12" src="../../../blog/2019/06/images/cancer_signature_figs/12.png"></p>
<ol>
<li>MUT_SEQの欠損値、chr23、24、25を除く。</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1">#mutのdfからchr23、24、MUT_SEQのnanを除く</span>
<span class="n">df_cut_chr</span><span class="o">=</span><span class="n">df_cut</span><span class="p">[(</span><span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;23&#39;</span><span class="p">)</span> <span class="o">&amp;</span><span class="n">amp</span><span class="p">;</span><span class="n">amp</span><span class="p">;</span> <span class="p">(</span><span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;24&#39;</span><span class="p">)</span> <span class="o">&amp;</span><span class="n">amp</span><span class="p">;</span><span class="n">amp</span><span class="p">;</span> <span class="p">(</span><span class="n">df_cut</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;25&#39;</span><span class="p">)]</span>
<span class="n">df_cut_chr</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;MUT_SEQ&#39;</span><span class="p">],</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="c1">#indexを振り直す</span>

<span class="n">df_cut_chr_i</span><span class="o">=</span><span class="n">df_cut_chr</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
</code></pre></div>


<p>なぜかCOSMICのデータにchr22以上のものが含まれていたので、除きました。</p>
<h4><span id="pelican-subsections-17">リファレンス配列から、MUT_SEQの前後の配列を抽出</span></h4>
<div class="highlight"><pre><span></span><code><span class="c1">#refから変異位置の前後の配列を抽出</span>
<span class="n">seq_before</span><span class="o">=</span><span class="p">[]</span>
<span class="n">seq_after</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="p">)):</span>
<span class="n">seq_before</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">list_seq</span><span class="p">[</span><span class="n">list_id</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">])][</span><span class="nb">int</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;single position&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">])</span><span class="o">-</span><span class="mi">2</span><span class="p">])</span>
<span class="n">seq_after</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">list_seq</span><span class="p">[</span><span class="n">list_id</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">])][</span><span class="nb">int</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;single position&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">])])</span>
</code></pre></div>


<h4><span id="pelican-subsections-18">COSMICのdfと合わせていく。</span></h4>
<ol>
<li>変異の前後の配列について、dfに追加し、大文字に直す</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_before&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">seq_before</span><span class="p">)</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_after&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">seq_after</span><span class="p">)</span><span class="c1">#大文字に変更</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_before&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_before&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_after&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_after&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
</code></pre></div>


<ol>
<li>前、変異、後の3つの配列を新しいカラムの中に追加する
（以下のようなややこしいカラムを作っているのは（＊注1）のためです）</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1">#3文字の配列にする（seq_forwardが鋳型）</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_forward&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_before&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;MUT_SEQ&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_after&#39;</span><span class="p">]</span><span class="c1">#鋳型鎖の順番を逆にしたもの（相補鎖を作るため）</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_forward_r&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_after&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;MUT_SEQ&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_before&#39;</span><span class="p">]</span>
</code></pre></div>


<p>ここまでで以下のようになります。（dfの右側だけ）</p>
<p><img alt="13" src="../../../blog/2019/06/images/cancer_signature_figs/13.png"></p>
<ol>
<li>3つの配列について、相補鎖を作ります</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1">#辞書型で相補鎖に変換</span>
<span class="n">dict_base</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span><span class="s2">&quot;T&quot;</span><span class="p">,</span><span class="s2">&quot;G&quot;</span><span class="p">:</span><span class="s2">&quot;C&quot;</span><span class="p">,</span><span class="s2">&quot;C&quot;</span><span class="p">:</span><span class="s2">&quot;G&quot;</span><span class="p">,</span><span class="s2">&quot;T&quot;</span><span class="p">:</span><span class="s2">&quot;A&quot;</span><span class="p">}</span>

<span class="n">list_seq_reverse</span><span class="o">=</span><span class="p">[]</span>
<span class="n">list_wt_seq_reverse</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="p">)):</span>
<span class="n">list_seq_reverse</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_forward_r&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="n">dict_base</span><span class="p">)))</span>
<span class="c1">#次のために、変異の配列も変換しておきます。 list_wt_seq_reverse.append((str(df_cut_chr_i[&#39;WT_SEQ&#39;][i])).translate(str.maketrans(dict_base)))</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;seq_reverse&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">list_seq_reverse</span><span class="p">)</span>
<span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;WT_seq_reverse&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">list_wt_seq_reverse</span><span class="p">)</span>
</code></pre></div>


<h4><span id="pelican-subsections-19">変異の情報をテキストファイルに書き出す</span></h4>
<p>いよいよ最後のステップです。</p>
<div class="highlight"><pre><span></span><code><span class="c1">#ID_tumorのカウント</span>
<span class="n">dict_tumor_id</span><span class="o">=</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;ID_tumour&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span><span class="n">list_tumor_id</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">dict_tumor_id</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</code></pre></div>


<p>まずはIDをリストに格納します。</p>
<div class="highlight"><pre><span></span><code><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;ID_tumour&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</code></pre></div>


<p><img alt="14" src="../../../blog/2019/06/images/cancer_signature_figs/14.png"></p>
<p>（上はテストファイルで行ったため、実際の数値とは異なります。）</p>
<p>いろいろな試行錯誤したコードで、テキストファイルに書き出していきます。</p>
<div class="highlight"><pre><span></span><code><span class="n">df_index_for_id</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">list_tumor_id</span><span class="p">:</span>
<span class="n">df_index_for_id</span><span class="p">[</span><span class="s1">&#39;index_</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">((</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="n">df_cut_chr_i</span><span class="p">[</span><span class="s1">&#39;ID_tumour&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">df_index_for_id</span><span class="p">[</span><span class="s1">&#39;index_</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)]:</span>
<span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">j</span><span class="p">)</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="k">continue</span>
<span class="k">else</span><span class="p">:</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;text_2/</span><span class="si">{}</span><span class="s2">.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="k">if</span> <span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;WT_SEQ&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;C&#39;</span><span class="ow">or</span> <span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;WT_SEQ&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;T&#39;</span><span class="p">:</span>
<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">((</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;WT_SEQ&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;&amp;amp;gt;&#39;</span><span class="o">+</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;MUT_SEQ&#39;</span><span class="p">]</span><span class="o">+</span>\
<span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;seq_forward&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span>\
<span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;seq_forward&#39;</span><span class="p">][</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
<span class="k">elif</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;WT_SEQ&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;A&#39;</span><span class="ow">or</span> <span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;WT_SEQ&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;G&#39;</span><span class="p">:</span>
<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">((</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;WT_seq_reverse&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;&amp;amp;gt;&#39;</span><span class="o">+</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;seq_reverse&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span>\
<span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;seq_reverse&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span>\
<span class="p">(</span><span class="n">df_cut_chr_i</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">j</span><span class="p">)])[</span><span class="s1">&#39;seq_reverse&#39;</span><span class="p">][</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
<span class="k">continue</span>
</code></pre></div>


<p>（特に速度のことは考えずに直感的にforで回したので、かなり時間がかかります。）
ここまでで変異のデータをIDごとに抽出することができました！</p>
<p>上のスクリプトをサーバーにでも投げてしばらく待ちます。</p>
<h3><span id="pelican-subsections-20">トピックモデルを使ってみる</span></h3>
<p><img alt="15" src="../../../blog/2019/06/images/cancer_signature_figs/15.png"></p>
<p>4までの操作で各IDごとの上のように、変異データが得られました。</p>
<p>これをトピックモデルを使って解析していきます。</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">gensim</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">wordcloud</span> <span class="kn">import</span> <span class="n">WordCloud</span>
</code></pre></div>


<h4><span id="pelican-subsections-21">ディレクトリ内のファイルリストを取得する</span></h4>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="s2">&quot;/Users/***/test/cancer_genome&quot;</span><span class="p">)</span>
<span class="n">list_file_name</span><span class="o">=</span><span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;./text_2/*&quot;</span><span class="p">)</span><span class="n">list_file_name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div>


<h4><span id="pelican-subsections-22">変異データを一つのリスト内に格納する</span></h4>
<div class="highlight"><pre><span></span><code><span class="n">list_text</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">list_file_name</span><span class="p">:</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">((</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="s1">&#39;r&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span><span class="o">!=</span><span class="mi">1</span><span class="p">:</span>
<span class="k">continue</span>
<span class="k">else</span><span class="p">:</span>
<span class="n">list_text</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">list</span><span class="p">(((</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="s1">&#39;r&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">readlines</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()))</span>
</code></pre></div>


<h4><span id="pelican-subsections-23">トピックモデル</span></h4>
<div class="highlight"><pre><span></span><code><span class="c1">#辞書を作成</span>
<span class="n">dictionary</span> <span class="o">=</span> <span class="n">gensim</span><span class="o">.</span><span class="n">corpora</span><span class="o">.</span><span class="n">Dictionary</span><span class="p">(</span><span class="n">list_text</span><span class="p">)</span>
<span class="n">corpus</span> <span class="o">=</span> <span class="p">[</span><span class="n">dictionary</span><span class="o">.</span><span class="n">doc2bow</span><span class="p">(</span><span class="n">text</span><span class="p">)</span> <span class="k">for</span> <span class="n">text</span> <span class="ow">in</span> <span class="n">list_text</span><span class="p">]</span><span class="c1">#トピックモデルの学習</span>
<span class="n">num_topics</span> <span class="o">=</span> <span class="mi">20</span>

<span class="n">lda</span> <span class="o">=</span> <span class="n">gensim</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">ldamodel</span><span class="o">.</span><span class="n">LdaModel</span><span class="p">(</span>
<span class="n">corpus</span><span class="o">=</span><span class="n">corpus</span><span class="p">,</span>
<span class="n">num_topics</span><span class="o">=</span><span class="n">num_topics</span><span class="p">,</span>
<span class="n">id2word</span><span class="o">=</span><span class="n">dictionary</span>
<span class="p">)</span>
</code></pre></div>


<p>トピック数（num_topics）については後述します。</p>
<p>これでトピックモデルの学習ができました。ここは実行時間はそこまで長くありません。</p>
<div class="highlight"><pre><span></span><code><span class="n">lda</span><span class="o">.</span><span class="n">show_topic</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
</code></pre></div>


<p><img alt="16" src="../../../blog/2019/06/images/cancer_signature_figs/16.png"></p>
<p>結果を見てみると、各トピックごとに、各変異がどれほどの割合を占めているのか、というのがわかります。</p>
<h4><span id="pelican-subsections-24">可視化</span></h4>
<p>おなじみmatplotlibを使ってグラフにしてみます。</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">collections</span>

<span class="n">clist</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">),</span> <span class="p">(</span><span class="mi">16</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">17</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;orange&quot;</span><span class="p">),(</span><span class="mi">32</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;orange&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">33</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;green&quot;</span><span class="p">),(</span><span class="mi">48</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;green&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">49</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;salmon&quot;</span><span class="p">),(</span><span class="mi">66</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;salmon&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">67</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">),(</span><span class="mi">83</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">84</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">),(</span><span class="mi">102</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">)]</span>
<span class="n">rvb</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">clist</span><span class="p">)</span>

<span class="n">n</span><span class="o">=</span><span class="mi">102</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">30</span><span class="p">))</span>
<span class="n">dict_topic</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">lda</span><span class="o">.</span><span class="n">show_topic</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">102</span><span class="p">))</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
<span class="n">x</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dict_topic</span><span class="p">)))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dict_topic</span><span class="p">)),</span> <span class="n">dict_topic</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">align</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">rvb</span><span class="p">(</span><span class="n">x</span><span class="o">/</span><span class="mi">102</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dict_topic</span><span class="p">)),</span> <span class="nb">list</span><span class="p">(</span><span class="n">dict_topic</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>


<p><img alt="17" src="../../../blog/2019/06/images/cancer_signature_figs/17.png"></p>
<p>リストで色を指定することで、このように変異の種類（T&gt;C、T&gt;A、C&gt;Gなど）ごとに色づけすることができました。
- 参考：<a href="https://stackoverflow.com/questions/42656585/barplot-colored-according-a-colormap">https://stackoverflow.com/questions/42656585/barplot-colored-according-a-colormap</a></p>
<p>これをトピックごとにグラフを作成し、画像として保存します。</p>
<div class="highlight"><pre><span></span><code><span class="n">clist</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;aqua&quot;</span><span class="p">),</span> <span class="p">(</span><span class="mi">16</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;aqua&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">17</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">),(</span><span class="mi">32</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">33</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">),(</span><span class="mi">48</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">49</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;gray&quot;</span><span class="p">),(</span><span class="mi">66</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;gray&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">67</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;lime&quot;</span><span class="p">),(</span><span class="mi">83</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;lime&quot;</span><span class="p">),</span>\
<span class="p">(</span><span class="mi">84</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;pink&quot;</span><span class="p">),(</span><span class="mi">102</span><span class="o">/</span><span class="mi">102</span><span class="p">,</span> <span class="s2">&quot;pink&quot;</span><span class="p">)]</span>
<span class="n">rvb</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">clist</span><span class="p">)</span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">lda</span><span class="o">.</span><span class="n">num_topics</span><span class="p">):</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">40</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
<span class="c1">#plt.subplot(lda.num_topics,2,i+1)</span>
<span class="n">dict_topic</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">lda</span><span class="o">.</span><span class="n">show_topic</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="mi">102</span><span class="p">))</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
<span class="n">x</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dict_topic</span><span class="p">)))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;topic #</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dict_topic</span><span class="p">)),</span> <span class="n">dict_topic</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">align</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">rvb</span><span class="p">(</span><span class="n">x</span><span class="o">/</span><span class="mi">102</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dict_topic</span><span class="p">)),</span> <span class="nb">list</span><span class="p">(</span><span class="n">dict_topic</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;topic_pdf/topicmodel_ms_</span><span class="si">{}</span><span class="s1">.pdf&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
</code></pre></div>


<p>（色はCOSMICのグラフのものと揃えました）</p>
<p><img alt="18" src="../../../blog/2019/06/images/cancer_signature_figs/18.png"></p>
<p><img alt="19" src="../../../blog/2019/06/images/cancer_signature_figs/19.png"></p>
<h3><span id="pelican-subsections-25">データベースのシグネチャと比較</span></h3>
<p>COSMICには、今までに推定されてきたがんゲノムの変異シグネチャーが多数登録されています。
- <a href="https://cancer.sanger.ac.uk/cosmic/signatures/SBS/">https://cancer.sanger.ac.uk/cosmic/signatures/SBS/</a></p>
<p><img alt="20" src="../../../blog/2019/06/images/cancer_signature_figs/20.png"></p>
<p>シグネチャーのグラフと、関連するがんの原因が示されています。</p>
<h4><span id="pelican-subsections-26">推定</span></h4>
<p>今回の課題は、「<strong>たばこ/紫外線のシグネチャーを見つけることができたか</strong>」でした。</p>
<p>データベースにあるグラフとどれほど一致しているか見ていきます。</p>
<p>COSMICに掲載されているシグネチャーSBS1〜SBS85までのうち、</p>
<ul>
<li><strong>たばこ</strong>（tobacco）が原因とされるもの→SBS4、5、29</li>
<li><strong>紫外線</strong>（ultraviolet light）が原因とされるもの→SBS7a〜d、38</li>
</ul>
<p><img alt="21" src="../../../blog/2019/06/images/cancer_signature_figs/21.png"></p>
<p>であるとわかりました。</p>
<h4><span id="pelican-subsections-27">推定結果</span></h4>
<p><strong>-たばこ-</strong></p>
<p>①SBS4</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="22" src="../../../blog/2019/06/images/cancer_signature_figs/22.png"></p>
<ul>
<li>推定</li>
</ul>
<p><img alt="23" src="../../../blog/2019/06/images/cancer_signature_figs/23.png"></p>
<p>②SBS5</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="24" src="../../../blog/2019/06/images/cancer_signature_figs/24.png"></p>
<ul>
<li>推定</li>
</ul>
<p><img alt="25" src="../../../blog/2019/06/images/cancer_signature_figs/25.png"></p>
<p>③SBS29</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="26" src="../../../blog/2019/06/images/cancer_signature_figs/26.png"></p>
<ul>
<li>推定</li>
</ul>
<p><img alt="27" src="../../../blog/2019/06/images/cancer_signature_figs/27.png"></p>
<p><strong>-紫外線-</strong></p>
<p>①SBS7a
- COSMIC</p>
<p><img alt="28" src="../../../blog/2019/06/images/cancer_signature_figs/28.png"></p>
<ul>
<li>推定</li>
</ul>
<p><img alt="29" src="../../../blog/2019/06/images/cancer_signature_figs/29.png"></p>
<p>②SBS7b</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="30" src="../../../blog/2019/06/images/cancer_signature_figs/30.png"></p>
<ul>
<li>推定</li>
</ul>
<p><img alt="31" src="../../../blog/2019/06/images/cancer_signature_figs/31.png"></p>
<p>③SBS7c</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="32" src="../../../blog/2019/06/images/cancer_signature_figs/32.png"></p>
<ul>
<li>推定</li>
</ul>
<p>類似しているシグネチャーは確認できませんでした。</p>
<p>④SBS7d</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="33" src="../../../blog/2019/06/images/cancer_signature_figs/33.png"></p>
<ul>
<li>推定</li>
</ul>
<p>類似しているシグネチャーは確認できませんでした。</p>
<p>⑤SBS38</p>
<ul>
<li>COSMIC</li>
</ul>
<p><img alt="34" src="../../../blog/2019/06/images/cancer_signature_figs/34.png"></p>
<ul>
<li>推定</li>
</ul>
<p>類似しているシグネチャーは確認できませんでした。</p>
<p>どれも完全に一致しているとは言えないのですが、近いものはそれなりに類似しているグラフが作れたのではないかと思います。</p>
<p>シグネチャーを見るだけでも、「たばこを原因とするがんは、変異の種類が多い」、「紫外線を原因とするがんは、変異の種類が少ない」という傾向も見て取れます。</p>
<h2><span id="pelican-subsections-28">おまけ -WordCloudによる可視化-</span></h2>
<p>トピックモデルで単語を扱う際によく同時にやられているものとして、WordCloudがあったため、変異データでもやってみました。</p>
<div class="highlight"><pre><span></span><code><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">30</span><span class="p">))</span>
<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">lda</span><span class="o">.</span><span class="n">num_topics</span><span class="p">):</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="n">t</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">lda</span><span class="o">.</span><span class="n">show_topic</span><span class="p">(</span><span class="n">t</span><span class="p">,</span><span class="mi">200</span><span class="p">))</span>
<span class="n">im</span> <span class="o">=</span> <span class="n">WordCloud</span><span class="p">()</span><span class="o">.</span><span class="n">generate_from_frequencies</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">im</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Topic #&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">))</span>
</code></pre></div>


<p><img alt="35" src="../../../blog/2019/06/images/cancer_signature_figs/35.png"></p>
<h2><span id="pelican-subsections-29">今後やってみたいこと</span></h2>
<blockquote>
<p>がんゲノムの変異シグネチャーは、最初に発表された論文ではNon- negative Matrix Factorization(NMF)(非負値行列因子分解)を使って推定された。2019年になって「トピックモデル」を使って変異 シグネチャを推定する論文が二報発表された。<br>
1. Funnell, T. et al. Integrated structural variation and point mutation signatures in cancer genomes using correlated topic models. PLOS Computational Biology 15, e1006799 (2019).
2. Matsutani, T., Ueno, Y., Fukunaga, T. &amp; Hamada, M. Discovering novel mutation signatures by latent Dirichlet allocation with variational Bayes inference. Bioinformatics doi:10.1093/bioinformatics/btz266</p>
</blockquote>
<p>このような背景があったため今回はトピックモデルを使ってみたのですが、NMFを用いた推定もやってみたいです。</p>
<p>変異シグネチャーの個数については、今回のセッティングが最適ではなく、丁寧にやるならhierarchical dirichlet process (HDP) を用いるか、perplexity、coherenceなどの評価指標を用いる必要があるのかと思いました。
- 参考：<a href="https://www.randpy.tokyo/entry/word2vec_skip_gram_model">https://www.randpy.tokyo/entry/word2vec_skip_gram_model</a></p>
<h2><span id="pelican-subsections-30">参考</span></h2>
<p>今回のコード、作成したグラフ等はGitHubにあります。
- <a href="https://github.com/ykohki/cancer_signature">https://github.com/ykohki/cancer_signature</a></p>
<p>トピックモデルを用いたがんゲノムの変異シグネチャー解析
- <a href="https://ipsj.ixsq.nii.ac.jp/ej/index.php?active_action=repository_view_main_item_detail&amp;page_id=13&amp;block_id=8&amp;item_id=182433&amp;item_no=1">https://ipsj.ixsq.nii.ac.jp/ej/index.php?active_action=repository_view_main_item_detail&amp;page_id=13&amp;block_id=8&amp;item_id=182433&amp;item_no=1</a></p>
<p>がんゲノムビッグデータから喫煙による遺伝子異常を同定
- <a href="https://www.ncc.go.jp/jp/information/pr_release/2016/1104/index.html">https://www.ncc.go.jp/jp/information/pr_release/2016/1104/index.html</a>
- <a href="https://science.sciencemag.org/content/354/6312/618">https://science.sciencemag.org/content/354/6312/618</a></p>
<p>同じ変異シグネチャーについて調べていて、特にたばこについて見ている論文でした
- https://www.nature.com/articles/nature12477#associating-cancer-aetiology-and-mutational-signatures</p>
  </div>
  
<div class="article-tag-list">
<span class="label label-default">Tags</span>
	<a href="../../../tag/bioinformatics.html"><i class="fa fa-tag"></i>Bioinformatics</a>&nbsp;
</div>  <hr />
  <div class="well well-sm">  <!-- Social media sharing buttons -->
    <!-- Twitter -->
    <a href="https://twitter.com/share" class="twitter-share-button" 
       data-via="oumed_python" >Tweet</a>&nbsp;

    <!-- Google+ -->
    <div class="g-plus" data-action="share" data-annotation="bubble"></div>
    &nbsp;&nbsp;&nbsp;&nbsp;
    <div class="g-plusone" data-size="medium"></div>&nbsp;

    <!-- Facebook -->
    <div class="fb-like" 
        data-href="../../../blog/2019/06/cancer_signature.html" 
        data-layout="button_count" 
        data-action="like" data-show-faces="true" 
        data-share="true">
    </div>
    &nbsp;
  </div> <!-- /Social media sharing buttons --><!-- Comment system with utteranc.es -->
<script src="https://utteranc.es/client.js"
        repo="oumpy/oumpy.github.io"
        issue-term="title"
        theme="github-light"
        crossorigin="anonymous"
        async>
</script><div class="related-posts">
    <h2>関連記事</h2>
    <ul>
        <li><a href="../../../blog/2019/12/bioinfoexam2019_answer.html">バイオインフォマティクス技術者認定試験2019 解答速報？</a></li>
        <li><a href="../../../blog/2019/03/velocyto.html">VELOCYTO</a></li>
        <li><a href="../../../blog/2019/03/original_rnaseq_pipeline.html">阪医Python会特製 RNA-seq pipeline ver. 1.0 リリース</a></li>
    </ul>
</div>

</article>
        </div><!-- /content -->

        <div class="col-md-3 sidebar-nav" id="sidebar">
<div class="row"><div class="col-md-12 col-sm-4 col-xs-12">
<form id="cse-search-box" action="../../../search.html">
<input type="hidden" name="ie" value="UTF-8">
<input type="text" name="q" placeholder="サイト内を検索" style="height: 1.7em; width: 82.5%;">
<input type="submit" value="&#xf002;" class="fas" style="height: 1.7em; width: 15%;">
</form>
<div class="sidebar_box">
<div class="box-title"><i class="fa fa-comment fa-fw fa-lg"></i> ソーシャル</div>
<div class="box-contents">
<ul class="list-unstyled social-links">
  <li><a href="https://twitter.com/oumed_python" target="_blank" title="Twitter">
    <span style="color:#1DA1F2; font-size:larger;"><i class="fab fa-twitter"></i><span>
  </a></li>
  <li><a href="mailto:handai.python@gmail.com" target="_blank" title="E-mail">
    <span style="color:#0078D4; font-size:larger;"><i class="far fa-envelope"></i><span>
  </a></li>
  <li><a href="https://github.com/oumpy" target="_blank" title="GitHub Organization">
    <span style="color:#211F1F; font-size:larger;"><i class="fab fa-github"></i><span>
  </a></li>
  <li><a href="https://www.youtube.com/channel/UCh1eAeDCpsZeOh0Z9paNfHQ" target="_blank" title="YouTube Channel">
    <span style="color:#c4302b; font-size:larger;"><i class="fab fa-youtube"></i><span>
  </a></li>
  <li><a href="https://oum-python.connpass.com" target="_blank" title="Connpass">
    <span style="color:#000000; font-size:normal;"><img width="35px" src="https://connpass.com/static/img/72_72.png" style="display: inline;"/><span>
  </a></li>
  <li><a href="../../../feeds/all.atom.xml" target="_blank" title="Atom Feed">
    <span style="color:#00008b; font-size:smaller;"><i class="fa fa-rss fa-fw fa-lg"></i><span>
  </a></li>
  <li><a href="../../../feeds/all.rss.xml" target="_blank" title="RSS Feed">
    <span style="color:#f26522; font-size:normal;"><i class="fas fa-rss-square fa-fw fa-lg"></i><span>
  </a></li>
</ul>
</div>
</div>
<div class="sidebar_box">
<div class="box-title"><i class="fas fa-bell"></i> 新着記事</div>
<div class="box-contents">
<div class="horizontal-scroll">
<table class="recent-posts" id="recentposts">
  <tr><td colspan="3"><span class="recent-posts-year">2020</span></td></tr>
  <tr>
    <td class="recent-posts-date">7/14</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../blog.html">Blog</a>]</span></td>
    <td><a href="../../../blog/2020/07/bloodsugar_simulation.html"><span class="recent-posts-title">血糖値の数理シミュレーション</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">7/6</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../news.html">News</a>]</span></td>
    <td><a href="../../../news/2020/07/officialize_accept.html"><span class="recent-posts-title">医学部公認団体として正式承認されました</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">7/3</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../blog.html">Blog</a>]</span></td>
    <td><a href="../../../blog/2020/07/pythonparty_and_python.html"><span class="recent-posts-title">Python会とPythonとPython会で使われているPython</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">7/1</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../blog.html">Blog</a>]</span></td>
    <td><a href="../../../blog/2020/07/samestats.html"><span class="recent-posts-title">同じ基本統計量・異なる散布図のデータセットを焼きなまし法で生成する</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">7/1</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../blog.html">Blog</a>]</span></td>
    <td><a href="../../../blog/2020/07/studymeeting2020_07.html"><span class="recent-posts-title">【勉強会資料 2020 第7回】機械学習実践編 : Kerasで作る深層学習画像分類器</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">6/27</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../blog.html">Blog</a>]</span></td>
    <td><a href="../../../blog/2020/06/competition_bfs.html"><span class="recent-posts-title">幅優先探索の話 (競プロの話)</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">6/25</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../news.html">News</a>]</span></td>
    <td><a href="../../../news/2020/06/formalname.html"><span class="recent-posts-title">会正式名称の制定等について</span></a></td>
  </tr>
  <tr>
    <td class="recent-posts-date">6/24</td>
    <td><span class="recent-posts-prefix-category">[<a href="../../../blog.html">Blog</a>]</span></td>
    <td><a href="../../../blog/2020/06/studymeeting2020_06.html"><span class="recent-posts-title">【勉強会資料 2020 第6回】 機械学習とCNN入門</span></a></td>
  </tr>
</table>
</div>
</div>
</div>
<div class="sidebar_box">
<div class="box-title"><i class="fa fa-tags fa-fw fa-lg"></i> タグ一覧</div>
<p class="tag-cloud">
    <span class="tag-1">
      <a href="../../../tag/news.html">
          <span class="badge badge-primary">7</span>
        News
      </a>
    </span>
    <span class="tag-1">
      <a href="../../../tag/bioinformatics.html">
          <span class="badge badge-primary">12</span>
        Bioinformatics
      </a>
    </span>
    <span class="tag-3">
      <a href="../../../tag/lun-wen-matome.html">
          <span class="badge badge-primary">2</span>
        論文まとめ
      </a>
    </span>
    <span class="tag-2">
      <a href="../../../tag/jing-ji-puroguramingu.html">
          <span class="badge badge-primary">5</span>
        競技プログラミング
      </a>
    </span>
    <span class="tag-4">
      <a href="../../../tag/lun-wen-guan-lian.html">
          <span class="badge badge-primary">1</span>
        論文関連
      </a>
    </span>
    <span class="tag-3">
      <a href="../../../tag/jian-ding-shi-yan.html">
          <span class="badge badge-primary">2</span>
        検定試験
      </a>
    </span>
    <span class="tag-3">
      <a href="../../../tag/data-science-competition.html">
          <span class="badge badge-primary">2</span>
        Data Science Competition
      </a>
    </span>
    <span class="tag-4">
      <a href="../../../tag/github.html">
          <span class="badge badge-primary">1</span>
        GitHub
      </a>
    </span>
    <span class="tag-0">
      <a href="../../../tag/python.html">
          <span class="badge badge-primary">17</span>
        Python
      </a>
    </span>
    <span class="tag-4">
      <a href="../../../tag/hai-wai-liu-xue.html">
          <span class="badge badge-primary">1</span>
        海外留学
      </a>
    </span>
    <span class="tag-4">
      <a href="../../../tag/simulation.html">
          <span class="badge badge-primary">1</span>
        Simulation
      </a>
    </span>
    <span class="tag-1">
      <a href="../../../tag/statistics.html">
          <span class="badge badge-primary">9</span>
        Statistics
      </a>
    </span>
    <span class="tag-2">
      <a href="../../../tag/shell-script.html">
          <span class="badge badge-primary">3</span>
        Shell script
      </a>
    </span>
    <span class="tag-4">
      <a href="../../../tag/unix.html">
          <span class="badge badge-primary">1</span>
        Unix
      </a>
    </span>
    <span class="tag-1">
      <a href="../../../tag/mian-qiang-hui.html">
          <span class="badge badge-primary">7</span>
        勉強会
      </a>
    </span>
    <span class="tag-3">
      <a href="../../../tag/hadouea.html">
          <span class="badge badge-primary">2</span>
        ハードウェア
      </a>
    </span>
    <span class="tag-1">
      <a href="../../../tag/machine-learning.html">
          <span class="badge badge-primary">15</span>
        Machine Learning
      </a>
    </span>
</p>
</div></div> <!-- col -->
<div class="col-md-12 col-sm-8 col-xs-12">
<div class="sidebar_box">
<div class="box-title"><i class="fab fa-twitter fa-fw fa-lg"></i> Timeline</div>
<div class="box-contents">
    <a class="twitter-timeline" data-height=600 href=https://twitter.com/oumed_python?ref_src=twsrc%5Etfw>Tweets by oumed_python</a>
    <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
</div>
</div>
</div></div>  <!-- row, col -->        </div><!--/sidebar -->
      </div><!--/row-->
    </div><!--/.container /#main-container -->

    <footer id="site-footer">
<!-- colophon  -->
<!-- Site built using Pelican.  Theme based on VoidyBootstrap. -->
<address id="site-colophon" style="padding-top:0rem;padding-bottom:0.03rem;">
    <div class="text-center">
    <p class="text-dark" style="margin:0px">
        © 2020 大阪大学医学部 情報医科学研究会 (Python会)
    </p>
    <p class="text-muted">
    Site built using <a href="http://getpelican.com/" target="_blank">Pelican</a>
    powered by <a href="https://www.python.org/">Python</a>
    &nbsp;&bull;&nbsp; Theme based on
    <a href="http://www.voidynullness.net/page/voidy-bootstrap-pelican-theme/"
       target="_blank">VoidyBootstrap</a> by 
    <a href="http://www.robertiwancz.com/"
       target="_blank">RKI</a>  
    </p>
    </div>
</address><!-- /colophon  --> 
      <!-- Site built using Pelican.  Theme based on VoidyBootstrap. -->
    </footer>


    <!-- javascript -->
   
    <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js"
            integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj"
            crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
            integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo"
            crossorigin="anonymous"></script>
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/js/bootstrap.min.js"
            integrity="sha384-OgVRvuATP1z7JjHLkuOU7Xw704+h835Lr+6QL9UvYjZE3Ipu6Tp75j7Bh/kR0JKI"
            crossorigin="anonymous"></script>
    <!-- https://github.com/dallaslu/bootstrap-4-multi-level-dropdown -->
    <script src="https://raw.githubusercontent.com/dallaslu/bootstrap-4-multi-level-dropdown/master/bootstrap4-dropdown-ml-hack-hover.js"></script>


<!-- Facebook -->
<div id="fb-root"></div>
<script>(function(d, s, id) {
  var js, fjs = d.getElementsByTagName(s)[0];
  if (d.getElementById(id)) return;
  js = d.createElement(s); js.id = id;
  js.src = "//connect.facebook.net/en_US/all.js#xfbml=1";
  fjs.parentNode.insertBefore(js, fjs);
}(document, 'script', 'facebook-jssdk'));
</script>

<!-- Twitter -->
<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>

<!-- Google+ -->
<!-- Synchronous 
<script type="text/javascript" src="https://apis.google.com/js/plusone.js"></script>
-->
<!-- Asynchronous -->
<script type="text/javascript">
  (function() {
    var po = document.createElement('script'); po.type = 'text/javascript'; po.async = true;
    po.src = 'https://apis.google.com/js/platform.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(po, s);
  })();
</script><script>
const colab_img_url = 'https://colab.research.google.com/assets/colab-badge.svg'
const github_img_url = 'https://img.shields.io/static/v1?label=&logo=github&message=View%20On%20GitHub&labelColor=black&color=gray'
const colab_img = '<img class="notebook-badge-image" src="' + colab_img_url + '" alt="Open In Colab" title="Open In Colab" width="119px"></img>';
const github_img = '<img class="notebook-badge-image" src="' + github_img_url + '" alt="View On GitHub" title="View On GitHub" width="124px"></img>';
var check_ipynb = document.getElementsByClassName('highlight hl-ipython3');
if (check_ipynb.length > 0) {
    var doc0 = document.getElementById("colablink");
    var article_date = document.getElementsByName('article:date')[0].content;
    var article_year = article_date.slice(0, 4);
    var article_month = article_date.slice(5, 7);
    var article_category = document.getElementsByName('article:category:slug')[0].content; 
    if (parseInt(article_month) <= 3) {
      article_year = article_year - 1;
    }
    var article_name = document.getElementsByName('article:slug')[0].content;
    var source_url = document.getElementsByName('source-repository')[0].content;
    if (source_url.endsWith('.git')) {
      source_url = source_url.slice(0, -4);
    } else if (url_base.endsWith('/')) {
      source_url = source_url.slice(0, -1);
    }
    var source_url_hierarchies = source_url.split('/');
    var repname = source_url_hierarchies.pop();
    var account = source_url_hierarchies.pop();
    var urlroot_colab = "https://colab.research.google.com/github/"+account+"/"+repname+"/blob/master/content/articles/";
    var urlroot_github = source_url + "/blob/master/content/articles/";
    var filepath = article_year+"sy/"+article_category+"/"+article_name+".ipynb"
    var colab_url = urlroot_colab + filepath;
    var github_url = urlroot_github + filepath;
    var site_url = document.getElementsByName('siteurl')[0].content;
    var button = '<a href='+github_url+'>'+github_img+'</a>'
               +'<a href='+colab_url+'>'+colab_img+'</a>';
    doc0.innerHTML= button;
}

$(function() {
  const myhost = $(location).attr('hostname'); /* hogehoge.github.io */
  /* in order to work only on github pages, add:
  if (! myhost.match(/^[^\.]+\.github\.io$/)) {
    return;
  }
  */
  const account = myhost.split('.')[0]; /* hogehoge */
  const repo = account + '/' + myhost + '/blob/master';
  const root = 'https://colab.research.google.com/github/' + repo;
  $('a').each(function() {
    const target = $(this).attr('href').replace('https://'+myhost, '');
    if ((target.startsWith('.') || target.startsWith('/')) && target.endsWith('.ipynb')) {
      let upper = root + $(location).attr('pathname');
      let upper_split = upper.split('/');
      let target_split = target.split('/');
      upper_split.pop();
      let head = true;
      while (true) {
        if (head && target_split[0] == '') {
          target_split.shift();
          upper_split = root.split('/');
          break;
        } else if (target_split[0] == '.') {
          target_split.shift();
          head = false;
        } else if (target_split[0] == '..') {
          target_split.shift();
          upper_split.pop();
          head = false;
        } else {
          break;
        }
      }
      colablink = upper_split.join('/') + '/' + target_split.join('/');
      $(this).after(' <a href='+colablink+'>'+colab_img+'</a>');
    }
  });
});
</script><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script>
!function(t){"use strict";var n=function(n){return this.each(function(){var e,i,a=t(this),o=a.data(),c=[a],r=this.tagName,d=0;e=t.extend({content:"body",headings:"h1,h2,h3"},{content:o.toc||void 0,headings:o.tocHeadings||void 0},n),i=e.headings.split(","),t(e.content).find(e.headings).attr("id",function(n,e){return e||function(t){0===t.length&&(t="?");for(var n=t.replace(/\s+/g,"_"),e="",i=1;null!==document.getElementById(n+e);)e="_"+i++;return n+e}(t(this).text())}).each(function(){var n=t(this),e=t.map(i,function(t,e){return n.is(t)?e:void 0})[0];if(e>d){var a=c[0].children("li:last")[0];a&&c.unshift(t("<"+r+"/>").appendTo(a))}else c.splice(0,Math.min(d-e,Math.max(c.length-1,0)));var o=n.text();o=o.replace("¶",""),t("<li/>").appendTo(c[0]).append(t("<a/>").text(o).attr("href","#"+n.attr("id"))),d=e})})},e=t.fn.toc;t.fn.toc=n,t.fn.toc.noConflict=function(){return t.fn.toc=e,this},t(function(){n.call(t("[data-toc]"))})}(window.jQuery);
</script>
<script>
$(function(){
  $('a[href*="#"]').on('click',function() {
    var container = $('.container');
    var add_move = container.height() + 50
    var speed = 400;
    var href= $(this).attr("href");
    var target = $(href == "#" || href == "" ? 'html' : href);
    var position = target.offset().top - add_move;
    $('body,html').animate({scrollTop:position}, speed, 'swing');
    return false;
  });
});
</script>
<script type="text/javascript">
    $("#toc").toc({headings: "h2,h3"});
    if( $('#toc').is(':empty') ) {
      $('.toc_box').remove();
    }
</script><p class="pagetop" id="pagetop"><a href="#wrap"></a></p>
<script>
$(document).ready(function() {
    var pagetop = $('#pagetop');
    $(window).scroll(function () {
        if ($(this).scrollTop() > 600) {
            pagetop.fadeIn();
        } else {
            pagetop.fadeOut();
            }
        });
        pagetop.click(function () {
            $('body, html').animate({ scrollTop: 0 }, 500);
                return false;
    });
});
</script><script>
$('.dropdown-toggle').click(function() {
    var location = $(this).attr('href');
    window.location.href = location;
    return false;
});
</script>  </body>
</html>